### PR TITLE
Suggesting additions/modifications for NOpenCL

### PR DESCRIPTION
**Pull request by SunsetQuest**

### CODE REVIEW

**Comment by sharwell** [comment]
Hi @SunsetQuest,

It's great to have another contributor here; welcome!

It's going to take a bit of time for me to review all of this. Keep in mind that this pull request is a request to merge your **master** branch into the **master** branch here. This means if you push new commits to your **master** branch before the pull request is merged (or closed), they will automatically show up as additions to the end of the list of commits you see above. If you decide to become more active on GitHub, you may find it convenient to create a new branch for each of your intended pull requests. This will allow you to simultaneously work on more than one item at a time without risk of changing pull requests you considered complete.

File: NOpenCL/CommandQueue.cs [hunk 1]
```diff
@@ -8,7 +8,10 @@ namespace NOpenCL
     using System;
     using System.ComponentModel;
     using NOpenCL.SafeHandles;
+    using System.Diagnostics;
+    using System.Runtime.InteropServices;
 
+    [DebuggerDisplay("Device = {Device.Name}")]
```
Comment (sharwell) [inline NOpenCL/CommandQueue.cs:14]: This seems like a fine improvement.
- :question: Have you tested this to make sure it behaves properly? I know it results in calling into native code from the debugger display, but that might not be a problem.
- :question: Are you comfortable sending this as a separate pull request?

File: NOpenCL/Device.cs [hunk 2]
```diff
@@ -10,14 +10,26 @@ namespace NOpenCL
     using System.ComponentModel;
     using NOpenCL.SafeHandles;
     using DeviceInfo = NOpenCL.UnsafeNativeMethods.DeviceInfo;
+    using System.Diagnostics;
 
+    [DebuggerDisplay("Name = {Name}")]
```
Comment (sharwell) [inline NOpenCL/Device.cs:15]: :bulb: See my note/questions on the previous use of `[DebuggerDisplay]`. If you make a separate pull request, this use of DebuggerDisplay would be appropriate to include at the same time as the other.

File: NOpenCL.Test/Intel/Optimization.cs [hunk 3]
```diff
@@ -102,16 +102,13 @@ public void TestOptimization(int taskSize, int localWorkSize, bool warming, bool
...
+                    Console.WriteLine("Maximum workgroup size for this kernel  {0}", workGroupSizeMaximum);
 
                     if (warming)
                     {
                         Console.Write("Warming up OpenCL execution...");
-                        using (commandQueue.EnqueueNDRangeKernel(kernel, new[] { (IntPtr)globalWorkSize }, autoGroupSize ? null : new[] { (IntPtr)localWorkSize }))
-                        {
-                        }
-
+                        commandQueue.EnqueueNDRangeKernel(kernel, new[] { (IntPtr)globalWorkSize }, autoGroupSize ? null : new[] { (IntPtr)localWorkSize });
```
Comment (sharwell) [inline NOpenCL.Test/Intel/Optimization.cs:111]: :exclamation: Not only can we not do this here, but people using this library must make sure and follow the same pattern. It's the downside of returning the event handle from each operation.

:bulb: I filed issue #3 to reconsider this pattern, since it's ugly and likely to lead to incorrect usage scenarios.

File: NOpenCL/CommandQueue.cs (hunk 1 above)
Comment (SunsetQuest) [inline NOpenCL/CommandQueue.cs:14]: I have tested it and it worked great for me. I did not see any side-effects when I using it.  It made it much easier to identify what the device is.  

I'll try and send this in a separate pull request.  I'm still learning Git.  Since this is its own item should it normally be in its own pull request? (vs sending one pull request with a boat load of items like I did)

Comment (sharwell) [inline NOpenCL/CommandQueue.cs:14]: @SunsetQuest It can go either way really. The advantage of sending single-topic (smaller) pull requests is it makes it easier to quickly merge items while other proposals are still under review. For example, based on your answer to my questions I would be ready to merge the debugger display improvements right now, but I can't because they were combined with other work that I haven't made decisions on yet.

Comment (SunsetQuest) [inline NOpenCL/CommandQueue.cs:14]: I see, that makes sense.  All changes are approved in one large chunk so small parts cannot be approved. Also, if there are several outstanding pull requests at the same time then they would be more difficult to merge. Thank you for your help.

Comment (sharwell) [inline NOpenCL/CommandQueue.cs:14]: That's correct. The merge problem varies in severity from project to project. In my case, I'm used to doing complex merges and have good software to help, so even when things conflict it's rarely a real problem. :smile:

**Comment by sharwell** [comment]
This is out of date, but I'm reviewing the concepts and filing issues for individual items so the code can be updated to address each concern.

**Comment by sharwell** [comment]
@SunsetQuest I didn't file an issue for commit 34e3496524b45d5055f502a43035275f69697233. If that's a change you are still interested in can you file an issue describing the current problems and what you're trying to accomplish?

**Comment by sharwell** [comment]
@SunsetQuest I didn't file an issue for commit 8bfb501f6c8820fbaaedbc51afe36187d81a24fe. If that's a change you are still interested in, can you file an issue for it?

**Comment by sharwell** [comment]
@SunsetQuest I didn't file an issue for commit https://github.com/tunnelvisionlabs/NOpenCL/pull/1/commits/7b4304cf207d9a9779ec107ea39e448aa28b61fa. If that's a change you are still interested in, can you file an issue for it?

**Comment by sharwell** [comment]
@SunsetQuest I didn't file issues for commits https://github.com/tunnelvisionlabs/NOpenCL/pull/1/commits/abc7302b79d77040d65a4f2ae843ee769718ea0e and https://github.com/tunnelvisionlabs/NOpenCL/pull/1/commits/7e5404cdeda2ea612d3ab99343155c5f3b5bc023. If these contain changes you are still interested in, can you file issues for them separately?
